Pipewire related segfault with coredump when turning TV on or changing HDMI input with Kodi 21

<!--- Please fill out this template to the best of your ability. You can always edit this issue once you have created it. -->
<!--- Read the following link before you create a new problem report: https://kodi.wiki/view/HOW-TO:Submit_a_bug_report  -->
## Bug report
### Describe the bug
Here is a clear and concise description of what the problem is:
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
<!--- A bug report that is not clear will be closed -->
<!--- Put your text below this line -->

I use an always-on PC as my VPN server and Kodi media box, plugged into my HDMI TV. My TV automatically turns off after a while. I auto-restart the PC every day, autologin into my user into a Cinnamon desktop, and run Kodi as an autostart application; note the TV should be off at this point, and there is no setup for Kodi to start a TV or whatnot. I use Pipewire as my sound server and am running Archlinux.

In this setup I've experienced Kodi crashing abruptly, seemingly randomly; I've narrowed it down to either turning the TV on, switching HDMI input (from my TV remote), but it also happens with no interaction from the TV. For example; I automatically restart my PC every day at 6.30am, and noticed Kodi would crash (from the logs) at 7.20am, well before anyone's touched the TV.

The logs show very little information. They've actually been misleading as the last lines I've usually seen related to an XRandR query so I've been misled into thinking it was a display server, and tried a flurry of solutions involving Kernel Parameters and EDID files. However, I've since found out that Kodi was throwing a core dump after a segfault. Here is a link to it: https://paste.kodi.tv/zayedaguya.

As we can see, it clearly relates to Pipewire. Doing some research on this error I've found a few things online, I put a few links below. From there, I tried the first solution I came by, which was running with the environment variable: `KODI_AE_SINK=ALSA` in my `.desktop` file, and the error's completely disappeared; most likely proving it has something to do with Pipewire.

I'm not 100% sure this is a Kodi issue or a Pipewire to be fair; but it's clear it's the `kodi` process that throws the coredump.

Anyway (probably) related issues:
* https://github.com/xbmc/xbmc/issues/22054
* https://github.com/xbmc/xbmc/issues/26212

## Expected Behavior
Here is a clear and concise description of what was expected to happen:
<!--- Tell us what should happen -->
<!--- Put your text below this line -->

Kodi does not crash randomly; or when turning my TV on; or when switching HDMI input on my TV.

## Actual Behavior
<!--- Tell us what happens instead -->
<!--- Put your text below this line -->

Kodi crashes almost all the time when doing any of the above. Note I say "almost all the time", because I can't recall if there's been any instance over the last 500 times where maybe it didn't crash.

## Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->
<!--- Put your text below this line -->

So what's definitely working for me is to run Kodi with the environment variable: `KODI_AE_SINK=ALSA`. However I wanted to raise as it LOOKED like a Kodi bug to me. If it can be addressed, I would be happy to remove my environment variable and rely on the vanilla package in the future; and also I'll be happy to NOT have to change my Sound Server stack from Pipewire on future machines I'll setup.

### To Reproduce
Steps to reproduce the behavior:
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
<!--- Put your text below this line -->

Using a Pipewire sound server, connect a PC to an HDMI TV. When the TV is off, leave Kodi 21 running for a while. It should crash after an hour or so. Alternatively, turn the TV on. Kodi will crash. Finally, switch HDMI inputs from the TV if possible. Kodi should also crash.

### Debuglog
<!--- Put your text below this line -->
<!--- A debuglog is always mandatory when creating an issue. Provide one! -->

The debuglog can be found here:

https://paste.kodi.tv/pudocepoyi.kodi

Oh, we can also see a bunch of `metadata.tvshows.themoviedb.org.python` errors in there but I _think_ it's safe to say they're a different error altogether; they'll be for another bug report.

### Stack trace

Please also find the stack trace from the coredump here; got it using a `coredumpctl info` command. Note how the timestamp of the last Kodi debug log lines match the timestamp for the coredump:

https://paste.kodi.tv/zayedaguya

## Additional context or screenshots (if appropriate)

Here is some additional context or explanation that might help:
<!--- How has this bug affected you? What were you trying to accomplish? -->
<!--- Put your text below this line -->

As I said, I need my PC to be always on as it serves as a VPN server among other things. I also need it to be connected to my HDMI TV so I can access Kodi and my media anytime I turn the TV on; so I need Kodi to be flexible and gracefully pick up a Pipewire sink when the TV is off or when I switch it on.

The `.desktop` file for Kodi has an `Exec=` directive of: `kodi`. That's the one I changed to: `Exec=env KODI_AE_SINK=ALSA kod`; which solved my issue.

I run Archlinux (current Linux version `6.12.9`) with Kernel Parameters:

```kernel parameters
quiet
loglevel=0
rd.udev.log_level=0
udev.log_level=0
rd.systemd.show_status=false
systemd.show_status=false
vt.global_cursor_default=0
rw
root=UUID=blah-blah-blah-blah-blah
resume=UUID=blah-blah-blah-blah-blah
systemd.machine_id=blah
```

My Sound Server stack is comprised of a bunch of Pipewire related packages;
coming straight from the Archwiki: https://wiki.archlinux.org/title/PipeWire:

```packages
lib32-libpipewire 1:1.2.7-1
lib32-pipewire 1:1.2.7-1
lib32-pipewire-jack 1:1.2.7-1
libpipewire 1:1.2.7-1
libwireplumber 0.5.7-1
pipewire 1:1.2.7-1
pipewire-alsa 1:1.2.7-1
pipewire-audio 1:1.2.7-1
pipewire-jack 1:1.2.7-1
pipewire-pulse 1:1.2.7-1
pipewire-zeroconf 1:1.2.7-1
wireplumber 0.5.7-1
```

### Your Environment
Used Operating system:
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- Put your text below this line. Checkboxes can easily be ticked once issue is created -->
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Archlinux rolling
 - Kodi version: 21.2

<!--- End of this issue -->
*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

Hi, so, sorry but I'm actually struggling to build Kodi from @neo1973's branch. It's very likely to be my noobility. 

I'm following the build guide here: https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md, I'm using Archlinux, I'm **not sure** I have all missing dependencies because the package list is Debian/Ubuntu based so I was hoping I'd just see errors popup and fix them when that happens, but I'm essentially stuck on the very first `cmake` command.

```sh
git clone https://github.com/neo1973/xbmc.git
cd xbmc
git fetch --all
git switch fix_26358
cd ..
mkdir xbmc-build
cd xbmc-build
cmake ../xbmc -DCMAKE_INSTALL_PREFIX=/usr/local -DCORE_PLATFORM_NAME=x11 -DAPP_RENDER_SYSTEM=gl
```

.. which yields the error: 

```
CMake Warning:
  Ignoring extra path from command line:

   "../xbmc"


CMake Error: The source directory "/home/adrien/Code/xbmc" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

Scanning the `man` page for `cmake` and doing quick search online didn't help. I can see loads of `CMakeLists.txt` files under `xbmc/` but they're all nested; and have never used `cmake` before so I'm a little at a loss. 

Cmake version: 

```sh
$ cmake --version
cmake version 3.31.6

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

Right, so I installed Java 21.0.6 and gave it another go but I just seem to hit a wall before that with something related to `spdlog` right at the end of the build. I tried both with building `spdlog` as per the guide:

```
sudo make -C tools/depends/target/spdlog PREFIX=/usr/local
```

.. and with the `spdlog` package from Archlinux; in both case the build fails with the error message:

```
build/network/httprequesthandler/python/network_httprequesthandlers_python.a(HTTPPythonWsgiInvoker.cpp.o):HTTPPythonWsgiInvoker.cpp:function CHTTPPythonWsgiInvoker::executeScript(_IO_FILE*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, _object*):(.text+0x3819): error: undefined reference to 'spdlog::details::log_msg::log_msg(spdlog::source_loc, fmt::v10::basic_string_view<char>, spdlog::level::level_enum, fmt::v10::basic_string_view<char>)'
build/network/httprequesthandler/python/network_httprequesthandlers_python.a(HTTPPythonWsgiInvoker.cpp.o):HTTPPythonWsgiInvoker.cpp:function CHTTPPythonWsgiInvoker::executeScript(_IO_FILE*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, _object*):(.text+0x388b): error: undefined reference to 'spdlog::details::log_msg::log_msg(spdlog::source_loc, fmt::v10::basic_string_view<char>, spdlog::level::level_enum, fmt::v10::basic_string_view<char>)'
build/network/httprequesthandler/python/network_httprequesthandlers_python.a(HTTPPythonWsgiInvoker.cpp.o):HTTPPythonWsgiInvoker.cpp:function CHTTPPythonWsgiInvoker::executeScript(_IO_FILE*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, _object*):(.text+0x3bd8): error: undefined reference to 'spdlog::details::log_msg::log_msg(spdlog::source_loc, fmt::v10::basic_string_view<char>, spdlog::level::level_enum, fmt::v10::basic_string_view<char>)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/kodi.dir/build.make:568: kodi-x11] Error 1
make[2]: Leaving directory '/home/adrien/Code/xbmc-fix_26358/build'
make[1]: *** [CMakeFiles/Makefile2:6897: CMakeFiles/kodi.dir/all] Error 2
make[1]: Leaving directory '/home/adrien/Code/xbmc-fix_26358/build'
make: *** [Makefile:146: all] Error 2
```

I'm guessing the Archlinux `spdlog` package install takes precedence over the self-built one though. I couldn't uninstall `spdlog` because it's a dependency of the `kodi` package and I'm a little concerned it'll break my current setup, so I'll need to think about it a little before I try that.

--- a/xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp
+++ b/xbmc/cores/AudioEngine/Sinks/pipewire/AESinkPipewire.cpp
@@ -302,6 +302,7 @@ void CAESinkPipewire::EnumerateDevicesEx(AEDeviceInfoList& list, bool force)
   list.emplace_back(defaultDevice);
 
   auto& registry = pipewire->GetRegistry();
+  std::lock_guard lg(registry);
 
   for (const auto& [id, node] : registry.GetNodes())
   {

--- a/xbmc/cores/AudioEngine/Sinks/pipewire/PipewireRegistry.cpp
+++ b/xbmc/cores/AudioEngine/Sinks/pipewire/PipewireRegistry.cpp
@@ -43,8 +43,6 @@ void CPipewireRegistry::OnGlobalAdded(void* userdata,
                                       uint32_t version,
                                       const struct spa_dict* props)
 {
-  auto& registry = *reinterpret_cast<CPipewireRegistry*>(userdata);
-
   if (strcmp(type, PW_TYPE_INTERFACE_Node) != 0)
     return;
 
@@ -55,6 +53,8 @@ void CPipewireRegistry::OnGlobalAdded(void* userdata,
   if (strcmp(mediaClass, "Audio/Sink") != 0)
     return;
 
+  auto& registry = *reinterpret_cast<CPipewireRegistry*>(userdata);
+  std::lock_guard lg(registry);
   auto& nodes = registry.GetNodes();
 
   nodes[id] = std::make_unique<CPipewireNode>(registry, id, type);
@@ -65,6 +65,7 @@ void CPipewireRegistry::OnGlobalAdded(void* userdata,
 void CPipewireRegistry::OnGlobalRemoved(void* userdata, uint32_t id)
 {
   auto& registry = *reinterpret_cast<CPipewireRegistry*>(userdata);
+  std::lock_guard lg(registry);
   auto& nodes = registry.GetNodes();
 
   auto it = nodes.find(id);

--- a/xbmc/cores/AudioEngine/Sinks/pipewire/PipewireRegistry.h
+++ b/xbmc/cores/AudioEngine/Sinks/pipewire/PipewireRegistry.h
@@ -8,6 +8,8 @@
 
 #pragma once
 
+#include "threads/CriticalSection.h"
+
 #include <map>
 #include <memory>
 #include <string>
@@ -35,6 +37,10 @@ public:
 
   std::map<uint32_t, std::unique_ptr<CPipewireNode>>& GetNodes() { return m_nodes; }
 
+  // C++ BasicLockable requirements
+  void lock() { m_lock.lock(); }
+  void unlock() { m_lock.unlock(); }
+
 private:
   static void OnGlobalAdded(void* userdata,
                             uint32_t id,
@@ -59,6 +65,8 @@ private:
   std::unique_ptr<pw_registry, PipewireRegistryDeleter> m_registry;
 
   std::map<uint32_t, std::unique_ptr<CPipewireNode>> m_nodes;
+
+  CCriticalSection m_lock;
 };
 
 } // namespace PIPEWIRE

